File: jakamai/jakamai/c/damesrv.c
/*
 * Código fuente Jakamai
 * (C) 2001 Juan J. Amor
 * Programa bajo la proteccion de GPL 2.0
 *
 **
 * $Id: damesrv.c,v 1.1 2001/08/01 21:40:19 jjamor Exp $
 **
 */

/* System includes */
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <unistd.h>

/* Own includes */
#include "config.h"

void main_loop()
{

  int counter,aux;
  char buffer[MAXURLLENGHT];
  char returnedurl[MAXURLLENGHT+MAXSERVERLENGHT+1];

  char localfile[MAXURLLENGHT];

  struct stat aux2;

  counter = 0;

  setbuf(stdout,NULL);

  while (fgets(buffer,MAXURLLENGHT,stdin)) {

    /* Erase the last character if end-of-line */
    aux = strlen(buffer);
    if (buffer[aux-1] == '\n')
      buffer[aux-1] = '\0';

    sprintf(localfile,"%s/%s",localpath,buffer);

    /* Returns a round-robin URL, if local file exists.
       If not, returns a local URL to produce an ERROR */
    if ( stat(localfile,&aux2) < 0 )
      sprintf(returnedurl,"%s/%s\n",mainserver,buffer);
    else {
      counter = ((counter+1) % MAXSERVERS);
      sprintf(returnedurl,"%s/%s\n",servers[counter],buffer);
    }

    fputs(returnedurl,stdout);

  }

  return;

}


int main()
{
  main_loop();

  exit(0);
}

File: jakamai/jakamai/perl/damesrv.pl
#!/usr/bin/perl
##
# Código fuente Jakamai
# Copyright (C) 2001 Hispalinux / Juan J. Amor / Ramon G. Camus
# Programa bajo la proteccion de GPL 2.0
##
# $Id: damesrv.pl,v 1.2 2001/07/17 13:26:50 jjamor Exp $
#
# Copiado de: damesrv.pl, v1.9 2001/05/23 19:10:42 jjamor Exp (web-lucas)
#
##

# srand (time^(($$ << 15) + 3 * $$ + 1));

use LWP::UserAgent;

### PARTE DE CONFIGURACION

# CONFIGURACION ACTUAL:
# 1/11 hispalinux.digitel.es
# 2/11 lucas.olea.org
# 2/11 polinux.upv.es
# 2/11 ulises.adi.uam.es
# 2/11 ftp.gul.uc3m.es
# 2/11 dinamica1.fciencias.unam.mx

# Lista de servidores replica
@servers = (
	    'http://lucas.olea.org',
	    'http://ulises.adi.uam.es/LuCAS',
            'http://www.polinux.upv.es/LuCAS',
	    'http://ulises.adi.uam.es/LuCAS',
            'http://ftp.gul.uc3m.es/mirrors/LuCAS',
            'http://www.polinux.upv.es/LuCAS',
            'http://lucas.olea.org',
	    'http://hispalinux.digitel.es/LuCAS',
            'http://dinamica1.fciencias.unam.mx/Linux/LuCAS',
            'http://ftp.gul.uc3m.es/mirrors/LuCAS',
            'http://dinamica1.fciencias.unam.mx/Linux/LuCAS'
	    );

# Numero de servidores replica
$numservers = scalar(@servers);

# URL del servidor principal (no virtual host)
$mainserver = "http://torresquevedo.hispalinux.es/LuCAS";

# Directorio fisico de los ficheros en el servidor principal
$dirlucas = "/var/www/LuCAS";

### FIN DE PARTE CONFIGURABLE



$ua = new LWP::UserAgent; 
$ua->agent("LuCAS-forwarder/0.5 " . $ua->agent) ; 

# $sel = int(rand($numservers));

$| = 1;

$cnt = 0; # Round-robin

while (<STDIN>) {

    # Existe realmente en lucas:

    #my $req = new HTTP::Request HEAD => "$mainserver/$_" ; 
    #my $res = $ua->request($req);

    #$_ = "index.html" if ($_ =~ /\/$/);

    chop; # Quitar el \n del final

    my $ficheroelegido = "$dirlucas/$_";

    if (-r $ficheroelegido) {

	# Si la URL es valida buscamos un mirror que pueda servirla
	my $vueltas = $numservers;
	do {
	    $cnt = (($cnt+1) % $numservers);
	    $server = "@servers[$cnt]/$_";
	    $req = new HTTP::Request HEAD => $server; 
	    $res = $ua->request($req);
	    $vueltas--;
	} until ( ($res->is_success) || ($vueltas == 0) );

    } else { 
    	# Ni siquiera existe en LuCAS. Saldrá la pagina de error:
	$server="$mainserver/$_";
    }

    # Devolvemos la URL finalmente elegida
    print "$server\n"; 
    
}

